src/DataStructures/LeviCivitaIterator.hpp

```diff
@@ -0,0 +1,134 @@
+  // Increment the current permutation
+  LeviCivitaIterator& operator++() noexcept {
+    ++permutation_;
+    if (permutation_ >= signs_.size()) {
+      valid_ = false;
+    }
+    return *this;
+  }
+
+  // Return the current (multi-index) Index, an ordered list of indices
+  // Note: here and throughout, use cpp17::array, which is constexpr,
+  // unlike std::array, so that I can constexpr signs_ and indexes_ below.
+  const cpp17::array<size_t, Dim>& operator()() const noexcept {
+    return indexes_[permutation_];
+  }
+
+  // Return a specific index from the current (multi-index) Index
+  const size_t& operator[](size_t i) { return indexes_[permutation_][i]; }
+
+  // Return the sign of the Levi-Civita symbol with these indices
+  int sign() const noexcept { return signs_[permutation_]; };
+
+ private:
+  // Initialize the indexes
+  static constexpr cpp17::array<cpp17::array<size_t, Dim>, factorial(Dim)>
+  indexes() noexcept {
+    cpp17::array<cpp17::array<size_t, Dim>, factorial(Dim)> indexes{};
+    cpp17::array<size_t, Dim> index{};
+    cpp2b::iota(index.begin(), index.end(), size_t(0));
+
+    indexes[0] = index;
+
+    // cpp20::next_permutation generates the different permuations of index
+    // loop over them to fill in the rest of the permutations in indexes
+    size_t permutation = 1;
+    while (cpp20::next_permutation(index.begin(), index.end())) {
+      indexes[permutation] = index;
+      ++permutation;
+    }
+    return indexes;
+  };
+
+  static constexpr cpp17::array<int, factorial(Dim)> signs() noexcept {
+    cpp17::array<int, factorial(Dim)> signs{};
+
+    // By construction, the sign of the first permutation is +1
+    signs[0] = 1;
+
+    // How do you know whether the corresponding Levi Civita symbol is
+    // +1 or -1? To find out, compute the number, in the factoradic number
+    // system, corresponding to each permutation, and sum the digits. If the sum
+    // is even (odd), the corresponding Levi-Civita symbol will be +1 (-1). This
+    // works because there are Dim! unique permutations of the Levi Civita
+    // symbol indices, so each one can be represented by a
+    // (`Dim`)-digit factorial-number-system number. For more on the
+    // factoradic number system, see
+    // https://en.wikipedia.org/wiki/Factorial_number_system
+    auto factoradic_counter = cpp17::array<size_t, Dim>();
+    for (size_t permutation = 1; permutation < factorial(Dim); ++permutation) {
+      for (size_t i = 0; i < Dim; ++i) {
+        factoradic_counter[i]++;
+        if (factoradic_counter[i] < i + 1) {
+          break;
+        } else {
+          factoradic_counter[i] = 0;
+        }
+      }
+      signs[permutation] =
+          (cpp2b::accumulate(factoradic_counter.begin(),
+                             factoradic_counter.end(), size_t(0)) %
+           2) == 0
+              ? 1
+              : -1;
+    }
+    return signs;
+  };
+
+  static constexpr cpp17::array<cpp17::array<size_t, Dim>, factorial(Dim)>
```

Comment:
Can you actually store a `std::array`? We've been trying to use the `cpp17` array only in implementation details. I realize the internals need a `cpp17::array`, I'm hoping the interface doesn't.

Reply:
I can't figure out how to do this while keeping this quantity a static constexpr; I keep getting errors about std::array not being constexpr. Do you have a solution? 

Not sure a solution is necessary, though: I did change operator() to return a std::array, so users of the LeviCivitaIterator class never deal with cpp17::array, as the only use of cpp17::array are in the private portion of the class.

Reply:
You will need to copy into a `cpp17::array<std::array<size_t, Dim>, factorial(Dim)>` at the end of the `indexes` function. I think that'll work.All you really want is the inner array to be a `std::array`, the outer doesn't matter.

Reply:
I've tried several variants of this, all fail to compile, complaining that something related to array is not constexpr. Here is the simplest I tried: instead of returning indexes, do this at the end of the indexes function:
~~~~
cpp17::array<std::array<size_t, Dim>, factorial(Dim)> indexes_to_return{};
for(size_t i = 0; i < factorial(Dim); ++i) {
    indexes_to_return[i] = static_cast<std::array<size_t, Dim>>(indexes[i]);
}
return indexes_to_return;
~~~~

Can you suggest a better way to do the copy that the compiler would accept as constexpr? 

Other than this, I think this PR is ready for you to look at again.